src/model/metric.py

```diff
@@ -47,3 +57,93 @@ def update(self, data, output):
 
     def finalize(self):
         return self.total_correct / self.total_number
+
+
+class FIDScoreOffline(BaseMetric):
+    """
+    Module calculating FID score by saving all images into temporary directory
+    """
+
+    def __init__(self, output_key, target_key, unnorm_mean=(0.5,), unnorm_std=(0.5,), nickname="FID_InceptionV3"):
+        super().__init__(output_key, target_key, nickname)
+        self.from_tensor_to_pil = transforms.Compose([
+            UnNormalize(unnorm_mean, unnorm_mean),
+            transforms.ToPILImage()
+        ])
+        self.tmp_gt_dir = tempfile.TemporaryDirectory(prefix='gt_')
+        self.tmp_out_dir = tempfile.TemporaryDirectory(prefix='out_')
+
+    def clear(self):
+        self.tmp_gt_dir.cleanup()
+        self.tmp_out_dir.cleanup()
+        self.tmp_gt_dir = tempfile.TemporaryDirectory(prefix='gt_')
+        self.tmp_out_dir = tempfile.TemporaryDirectory(prefix='out_')
+
+    def _save_img_tensor(self, tensor, buffer_dir):
+        """ Save image tensor to a named temporary file and return the name."""
+        temp_f = tempfile.NamedTemporaryFile(suffix='.png', dir=buffer_dir.name, delete=False)
+        pil_image = self.from_tensor_to_pil(tensor.cpu())
+        pil_image.save(temp_f)
+        temp_f.close()
+
+    def update(self, data, output):
+        for gt_tensor, out_tensor in zip(data[self.target_key], output[self.output_key]):
+            self._save_img_tensor(gt_tensor, self.tmp_gt_dir)
+            self._save_img_tensor(out_tensor.clamp(-1, 1), self.tmp_out_dir)
+        return None
+
+    def finalize(self):
+        return fid_score.calculate_fid_given_paths(
+            paths=[self.tmp_gt_dir.name, self.tmp_out_dir.name],
+            batch_size=10, cuda=True, dims=2048)
+
+
+class FIDScoreOnline(BaseMetric):
+    """
+    Module calculating FID score online (store inception activation in memory)
+    """
+
+    def __init__(self, output_key, target_key, unnorm_mean=(0.5,), unnorm_std=(0.5,), nickname="FID_InceptionV3"):
+        super().__init__(output_key, target_key, nickname)
+        self._unNormalizer = UnNormalize(unnorm_mean, unnorm_mean)
+        block_idx = inception.InceptionV3.BLOCK_INDEX_BY_DIM[2048]
```

Comment:
Is FID bound with InceptionV3?

Reply:
No. Are you thinking that the model should be constructed outside `FIDScoreOnline`?

Reply:
It should probably be configurable or named as InceptionV3FIDScoreOnline?

Reply:
I've made it an abstract class such that it supports different backbone